README.md
# Cryptography Test Vectors

[![crates.io](https://img.shields.io/crates/v/crypto_vectors)](https://crates.io/crates/crypto_vectors)
[![docs](https://img.shields.io/docsrs/crypto_vectors)](https://docs.rs/crypto_vectors)
[![build & test](https://github.com/sheroz/crypto_vectors/actions/workflows/ci.yml/badge.svg)](https://github.com/sheroz/crypto_vectors/actions/workflows/ci.yml)
[![MIT](https://img.shields.io/github/license/sheroz/crypto_vectors)](https://github.com/sheroz/magma/tree/main/crypto_vectors/LICENSE.txt)

## Block Ciphers

### Test vectors for GOST ciphers

* [RFC 8891](https://datatracker.ietf.org/doc/html/rfc8891.html) a.k.a GOST R 34.12-2015: Block Cipher "Magma"
* [RFC 5830](https://datatracker.ietf.org/doc/html/rfc5830) a.k.a GOST 28147-89
* Block Cipher Modes:
  * [GOST R 34.13-2015](https://www.tc26.ru/standard/gost/GOST_R_3413-2015.pdf)
  * CTR-ACPKM: [RFC8645](https://www.rfc-editor.org/rfc/rfc8645.html#section-6.2.2), [Р 1323565.1.017—2018](https://standartgost.ru/g/%D0%A0_1323565.1.017-2018)

Link to [documentation](https://docs.rs/crypto_vectors/latest/crypto_vectors/gost/index.html)

---


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.1.4 (2023-07-27)

* updated documentation


src/gost/rfc5831.rs
//! Test vectors for [RFC5830, a.k.a GOST 28147-89](https://www.rfc-editor.org/rfc/rfc5830)
//! 
//! Test vectors taken from [RFC5831](https://www.rfc-editor.org/rfc/rfc5831#section-7)
//! 
//! **Cipher Keys**
//! 
//! K1 (little-endian)  = 0x733D2C20 0x65686573 0x74746769 0x79676120 0x626E7373 0x20657369 0x326C6568 0x33206D54
//! 
//! K1 (big-endian) = [0x33206D54, 0x326C6568, 0x20657369, 0x626E7373, 0x79676120, 0x74746769, 0x65686573, 0x733D2C20]    
//! 
//! K2 (little-endian)  = 0x110C733D 0x0D166568 0x130E7474 0x06417967 0x1D00626E 0x161A2065 0x090D326C 0x4D393320
//! 
//! K2 (big-endian) = [0x4D393320, 0x090D326C, 0x161A2065, 0x1D00626E, 0x06417967, 0x130E7474, 0x0D166568, 0x110C733D]    
//! 
//! K3 (little-endian)  = 0x80B111F3 0x730DF216 0x850013F1 0xC7E1F941 0x620C1DFF 0x3ABAE91A 0x3FA109F2 0xF513B239
//! 
//! K3 (big-endian) = [0xF513B239, 0x3FA109F2, 0x3ABAE91A, 0x620C1DFF, 0xC7E1F941, 0x850013F1, 0x730DF216, 0x80B111F3]
//! 
//! K4 (little-endian)  = 0xA0E2804E 0xFF1B73F2 0xECE27A00 0xE7B8C7E1 0xEE1D620C 0xAC0CC5BA 0xA804C05E 0xA18B0AEC
//! 
//! K4 (big-endian) = [0xA18B0AEC, 0xA804C05E, 0xAC0CC5BA, 0xEE1D620C, 0xE7B8C7E1, 0xECE27A00, 0xFF1B73F2, 0xA0E2804E]
//! 
//! **Plaintext**
//! 
//! 0x0
//! 
//! 
//! **Outputs**
//! 
//! S1 = 0x42ABBCCE 0x32BC0B1B
//! 
//! S2 = 0x5203EBC8 0x5D9BCFFD
//! 
//! S3 = 0x8D345899 0x00FF0E28
//! 
//! S4 = 0xE7860419 0x0D2A562D

/// Cipher Key K1
pub const CIPHER_KEY1: [u32;8] = [0x33206D54, 0x326C6568, 0x20657369, 0x626E7373, 0x79676120, 0x74746769, 0x65686573, 0x733D2C20];
/// Cipher Key K2
pub const CIPHER_KEY2: [u32;8] = [0x4D393320, 0x090D326C, 0x161A2065, 0x1D00626E, 0x06417967, 0x130E7474, 0x0D166568, 0x110C733D];    
/// Cipher Key K3
pub const CIPHER_KEY3: [u32;8] = [0xF513B239, 0x3FA109F2, 0x3ABAE91A, 0x620C1DFF, 0xC7E1F941, 0x850013F1, 0x730DF216, 0x80B111F3];
/// Cipher Key K4
pub const CIPHER_KEY4: [u32;8] = [0xA18B0AEC, 0xA804C05E, 0xAC0CC5BA, 0xEE1D620C, 0xE7B8C7E1, 0xECE27A00, 0xFF1B73F2, 0xA0E2804E];

/// Plaintext
pub const PLAINTEXT: u64 = 0;

/// Ciphertext1
pub const CIPHERTEXT1: u64 = 0x42ABBCCE32BC0B1B;
/// Ciphertext2
pub const CIPHERTEXT2: u64 = 0x5203EBC85D9BCFFD;
/// Ciphertext3
pub const CIPHERTEXT3: u64 = 0x8D34589900FF0E28;
/// Ciphertext4
pub const CIPHERTEXT4: u64 = 0xE78604190D2A562D;


src/gost/r3413_2015.rs
//! Test vectors for [Block Cipher Modes, GOST R 34.13-2015](https://www.tc26.ru/standard/gost/GOST_R_3413-2015.pdf)  
//! 
//! * **ECB** - Electronic Codebook Mode
//! * **CTR** - Counter Encryption Mode
//! * **OFB** - Output Feedback Mode
//! * **CBC** - Cipher Block Chaining Mode
//! * **CFB** - Cipher Feedback Mode
//! * **MAC** - Message Authentication Code Generation Mode

/// Cipher Key, Page 35, Section: A.2
pub const CIPHER_KEY: [u32; 8] = [
    0xffeeddcc, 0xbbaa9988, 0x77665544, 0x33221100, 0xf0f1f2f3, 0xf4f5f6f7, 0xf8f9fafb, 0xfcfdfeff,
];

/// Plaintext1, Page 35, Section: A.2
pub const PLAINTEXT1: u64 = 0x92def06b3c130a59_u64;
/// Plaintext2, Page 35, Section: A.2
pub const PLAINTEXT2: u64 = 0xdb54c704f8189d20_u64;
/// Plaintext3, Page 35, Section: A.2
pub const PLAINTEXT3: u64 = 0x4a98fb2e67a8024c_u64;
/// Plaintext4, Page 35, Section: A.2
pub const PLAINTEXT4: u64 = 0x8912409b17b57e41_u64;

/// Ciphertext1 in ECB Mode, Page 35, Section: A.2.1
pub const CIPHERTEXT1_ECB: u64 = 0x2b073f0494f372a0_u64;
/// Ciphertext2 in ECB Mode, Page 35, Section: A.2.1
pub const CIPHERTEXT2_ECB: u64 = 0xde70e715d3556e48_u64;
/// Ciphertext3 in ECB Mode, Page 35, Section: A.2.1
pub const CIPHERTEXT3_ECB: u64 = 0x11d8d9e9eacfbc1e_u64;
/// Ciphertext4 in ECB Mode, Page 35, Section: A.2.1
pub const CIPHERTEXT4_ECB: u64 = 0x7c68260996c67efb_u64;

/// Ciphertext1 in CTR Mode, Page 36, Section A.2.2
pub const CIPHERTEXT1_CTR: u64 = 0x4e98110c97b7b93c_u64;
/// Ciphertext2 in CTR Mode, Page 36, Section A.2.2
pub const CIPHERTEXT2_CTR: u64 = 0x3e250d93d6e85d69_u64;
/// Ciphertext3 in CTR Mode, Page 36, Section A.2.2
pub const CIPHERTEXT3_CTR: u64 = 0x136d868807b2dbef_u64;
/// Ciphertext4 in CTR Mode, Page 36, Section A.2.2
pub const CIPHERTEXT4_CTR: u64 = 0x568eb680ab52a12d_u64;

/// Ciphertext1 in OFB Mode, Page 37, Section A.2.3
pub const CIPHERTEXT1_OFB: u64 = 0xdb37e0e266903c83_u64;
/// Ciphertext2 in OFB Mode, Page 37, Section A.2.3
pub const CIPHERTEXT2_OFB: u64 = 0x0d46644c1f9a089c_u64;
/// Ciphertext3 in OFB Mode, Page 37, Section A.2.3
pub const CIPHERTEXT3_OFB: u64 = 0xa0f83062430e327e_u64;
/// Ciphertext4 in OFB Mode, Page 37, Section A.2.3
pub const CIPHERTEXT4_OFB: u64 = 0xc824efb8bd4fdb05_u64;

/// Ciphertext1 in CBC Mode, Page 38, Section A.2.4
pub const CIPHERTEXT1_CBC: u64 = 0x96d1b05eea683919_u64;
/// Ciphertext2 in CBC Mode, Page 38, Section A.2.4
pub const CIPHERTEXT2_CBC: u64 = 0xaff76129abb937b9_u64;
/// Ciphertext3 in CBC Mode, Page 38, Section A.2.4
pub const CIPHERTEXT3_CBC: u64 = 0x5058b4a1c4bc0019_u64;
/// Ciphertext4 in CBC Mode, Page 38, Section A.2.4
pub const CIPHERTEXT4_CBC: u64 = 0x20b78b1a7cd7e667_u64;

/// Ciphertext1 in CFB Mode, Page 39, Section A.2.5
pub const CIPHERTEXT1_CFB: u64 = 0xdb37e0e266903c83_u64;
/// Ciphertext2 in CFB Mode, Page 39, Section A.2.5
pub const CIPHERTEXT2_CFB: u64 = 0x0d46644c1f9a089c_u64;
/// Ciphertext3 in CFB Mode, Page 39, Section A.2.5
pub const CIPHERTEXT3_CFB: u64 = 0x24bdd2035315d38b_u64;
/// Ciphertext4 in CFB Mode, Page 39, Section A.2.5
pub const CIPHERTEXT4_CFB: u64 = 0xbcc0321421075505_u64;

/// Generating MAC, Page 40, Section A.2.6
pub const MAC: u32 = 0x154e7210_u32;


src/gost/r1323565_1_017_2018/ctr_acpkm.rs
//! Test Vectors for [RFC8645, CTR-ACPKM](https://www.rfc-editor.org/rfc/rfc8645.html#section-6.2.2) 
//! 
//! Test Vectors taken from [Р 1323565.1.017—2018](https://standartgost.ru/g/%D0%A0_1323565.1.017-2018),
//! Page 11,
//! Section A.1

/// Cipher Key
pub const CIPHER_KEY: [u8;32] = [
    0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF, 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77,
    0xFE, 0xDC, 0xBA, 0x98, 0x76, 0x54, 0x32, 0x10, 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF
];

/// Plaintext message
pub const PLAINTEXT: [u8;56] = [
    0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x00, 0xFF, 0xEE, 0xDD, 0xCC, 0xBB, 0xAA, 0x99, 0x88,
    0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xEE, 0xFF, 0x0A,
    0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xEE, 0xFF, 0x0A, 0x00,
    0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99            
];

/// Ciphertext
pub const CIPHERTEXT: [u8;56] = [
    0x2A, 0xB8, 0x1D, 0xEE, 0xEB, 0x1E, 0x4C, 0xAB, 0x68, 0xE1, 0x04, 0xC4, 0xBD, 0x6B, 0x94, 0xEA,
    0xC7, 0x2C, 0x67, 0xAF, 0x6C, 0x2E, 0x5B, 0x6B, 0x0E, 0xAF, 0xB6, 0x17, 0x70, 0xF1, 0xB3, 0x2E,
    0xA1, 0xAE, 0x71, 0x14, 0x9E, 0xED, 0x13, 0x82, 0xAB, 0xD4, 0x67, 0x18, 0x06, 0x72, 0xEC, 0x6F,
    0x84, 0xA2, 0xF1, 0x5B, 0x3F, 0xCA, 0x72, 0xC1
];


src/gost/rfc8891.rs
//! Test vectors for [RFC8891, a.k.a GOST R 34.12-2015: Block Cipher "Magma", Appendix A.](https://datatracker.ietf.org/doc/html/rfc8891.html#section-appendix.a)

/// [Section A.1. Transformation t](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.1)
pub const TRANSFORMATION_T: [(u32, u32); 4] = [
    (0xfdb97531, 0x2a196f34),
    (0x2a196f34, 0xebd9f03a),
    (0xebd9f03a, 0xb039bb3d),
    (0xb039bb3d, 0x68695433),
];

/// [Section A.2. Transformation g](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.2)
pub const TRANSFORMATION_G: [((u32, u32), u32); 4] = [
    ((0x87654321, 0xfedcba98), 0xfdcbc20c),
    ((0xfdcbc20c, 0x87654321), 0x7e791a4b),
    ((0x7e791a4b, 0xfdcbc20c), 0xc76549ec),
    ((0xc76549ec, 0x7e791a4b), 0x9791c849),
];

/// [Section A.3. Key Schedule](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.3)
pub const CIPHER_KEY: [u32; 8] = [
    0xffeeddcc, 0xbbaa9988, 0x77665544, 0x33221100, 0xf0f1f2f3, 0xf4f5f6f7, 0xf8f9fafb, 0xfcfdfeff,
];

/// [Section A.3. Key Schedule](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.3)
pub const CIPHER_KEY_U8_ARRAY: [u8; 32] = [
    0xff, 0xee, 0xdd, 0xcc, 0xbb, 0xaa, 0x99, 0x88, 0x77, 0x66, 0x55, 0x44, 0x33, 0x22, 0x11, 0x00,
    0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9, 0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff,
];

/// [Section A.3. Key Schedule](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.3)
pub const ROUND_KEYS: [u32; 32] = [
    0xffeeddcc, 0xbbaa9988, 0x77665544, 0x33221100, 0xf0f1f2f3, 0xf4f5f6f7, 0xf8f9fafb, 0xfcfdfeff,
    0xffeeddcc, 0xbbaa9988, 0x77665544, 0x33221100, 0xf0f1f2f3, 0xf4f5f6f7, 0xf8f9fafb, 0xfcfdfeff,
    0xffeeddcc, 0xbbaa9988, 0x77665544, 0x33221100, 0xf0f1f2f3, 0xf4f5f6f7, 0xf8f9fafb, 0xfcfdfeff,
    0xfcfdfeff, 0xf8f9fafb, 0xf4f5f6f7, 0xf0f1f2f3, 0x33221100, 0x77665544, 0xbbaa9988, 0xffeeddcc,
];

/// [Section A.4. Test Encryption](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.4)
pub const TRANSFORMATION_BIG_G: [(u32, u32); 32] = [
    (0x76543210, 0x28da3b14),
    (0x28da3b14, 0xb14337a5),
    (0xb14337a5, 0x633a7c68),
    (0x633a7c68, 0xea89c02c),
    (0xea89c02c, 0x11fe726d),
    (0x11fe726d, 0xad0310a4),
    (0xad0310a4, 0x37d97f25),
    (0x37d97f25, 0x46324615),
    (0x46324615, 0xce995f2a),
    (0xce995f2a, 0x93c1f449),
    (0x93c1f449, 0x4811c7ad),
    (0x4811c7ad, 0xc4b3edca),
    (0xc4b3edca, 0x44ca5ce1),
    (0x44ca5ce1, 0xfef51b68),
    (0xfef51b68, 0x2098cd86),
    (0x2098cd86, 0x4f15b0bb),
    (0x4f15b0bb, 0xe32805bc),
    (0xe32805bc, 0xe7116722),
    (0xe7116722, 0x89cadf21),
    (0x89cadf21, 0xbac8444d),
    (0xbac8444d, 0x11263a21),
    (0x11263a21, 0x625434c3),
    (0x625434c3, 0x8025c0a5),
    (0x8025c0a5, 0xb0d66514),
    (0xb0d66514, 0x47b1d5f4),
    (0x47b1d5f4, 0xc78e6d50),
    (0xc78e6d50, 0x80251e99),
    (0x80251e99, 0x2b96eca6),
    (0x2b96eca6, 0x05ef4401),
    (0x05ef4401, 0x239a4577),
    (0x239a4577, 0xc2d8ca3d),
    (0xc2d8ca3d, 0x4ee901e5),
];

/// [Section A.4. Test Encryption](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.4)
pub const PLAINTEXT: u64 = 0xfedcba9876543210_u64;

/// [Section A.4. Test Encryption](https://datatracker.ietf.org/doc/html/rfc8891.html#section-a.4)
pub const CIPHERTEXT: u64 = 0x4ee901e5c2d8ca3d_u64;
